### ISSUE TITLE
Java TurboVNC Viewer on Windows closes on Alt-F4

### ISSUE THREAD

**Issue by capi** [TurboVNC/turbovnc#434]
When pressing Alt-F4 the key-codes are sent to the VNC server, but the viewer also closes. When re-connecting, I can see that the Alt-F4 has also been sent to the server.

This happens, regardless if I use "Grab Keyboard" or not, even in Full-Screen mode.

Please let me know if there are any more information I can provide.

**Comment by dcommander** [comment]
Unfortunately Alt-F4 is intercepted by the operating system before our low-level keyboard hook can intercept it, so it isn't possible to grab that keystroke.

**Comment by capi** [comment]
Ok, sad to hear. Normally is possible in Java Programs to prevent Alt-F4 to close the Window, in Swing it's normally done by intercepting the window-closing event (I don't know how it is handled here).

As I said: the key-stroke is recognized and then sent to the server, all that is missing is preventing the local Window from being closed, and that's normally possible in all the GUI frameworks around.

I'll stick with 2.2.x then for some time longer :-)

**Comment by dcommander** [comment]
So the 2.2.x viewer (presumably the Windows native viewer) does not exhibit the issue?  That's odd, since it used the same low-level keyboard hook.

**Comment by dcommander** [comment]
As far as intercepting the window close event in Swing, it does look like there may be a way to tell it not to close the window, but I still want to understand why it works in TurboVNC 2.2.x.  (The 2.2.x viewer code did nothing special in this regard.)  I will do some more investigating.

**Comment by capi** [comment]
It works with the native viewer in 2.2.x, I'll need to double-check if it does with the Java version.

**Comment by capi** [comment]
Wow, thanks for the change! Looking forward to trying it out!
Do you really think that this should only be for grabbed state, though? In the 2.2.x branch grabbing was only needed to prevent Alt-Tab, but not Alt-F4.

**Comment by dcommander** [comment]
Yes, primarily for consistency with the Un*x viewer, which also only passes Alt-F4 if keyboard grabbing is enabled.  It's also consistent with the fact that other special keystrokes aren't passed unless keyboard grabbing is enabled.  The 2.2.x viewer behavior was more or less accidental.  I'm still not sure why Alt-F4 was disabled with it.  I am sensitive to the fact that users may be accustomed to that accidental behavior, but I am also sensitive to the fact that other users may be accustomed to using Alt-F4 to close the viewer window and disconnect.  I think it's good to support both.  I would be open to an undocumented system property (which could be set in %USERPROFILE%\.vnc\default.turbovnc) that causes Alt-F4 to always be sent, regardless of grabbing state.

**Comment by capi** [comment]
I understand your logic, it's sound. The undocumented property would be a great compromise. Thanks for considering it.
